source/common/config/grpc_mux_impl.cc

```diff
@@ -206,5 +220,47 @@ void GrpcMuxImpl::handleEstablishmentFailure() {
   }
 }
 
+void GrpcMuxImpl::populateControlPlaneInfo(const envoy::api::v2::DiscoveryResponse& message) {
+  if (message.has_control_plane()) {
+    Protobuf::util::MessageDifferencer message_differencer;
+    envoy::admin::v2alpha::ControlPlaneConfigDump::ConfigSourceControlPlaneInfo
+        config_source_control_plane_info;
+    bool config_exists = false;
+    for (auto& config_control_plane :
+         GrpcMuxImpl::per_service_control_plane_info_[server_type_url_]) {
+      if (message_differencer.Compare(grpc_service_, config_control_plane.grpc_service())) {
+        config_source_control_plane_info = config_control_plane;
+        config_exists = true;
+        break;
+      }
+    }
+    config_source_control_plane_info.mutable_control_plane()->set_identifier(
+        message.control_plane().identifier());
+    TimestampUtil::systemClockToTimestamp(
+        time_source_.systemTime(), *(config_source_control_plane_info.mutable_last_updated()));
+    config_source_control_plane_info.mutable_grpc_service()->MergeFrom(grpc_service_);
+    if (!config_exists) {
+      GrpcMuxImpl::per_service_control_plane_info_[server_type_url_].push_back(
+          config_source_control_plane_info);
+    }
+  }
```

Comment:
Is it possible that for the same service, e.g. EDS, we might have two identical `GrpcService` configurations, but they connect to distinct servers which return distinct identifiers in their `DiscoveryResponse`?

Reply:
Yes. It is possible and in that case, it goes in to update mode and updates the control_plane_identifier for existing gRPC service.

Reply:
if cluster X points to GrpcService M for EDS and cluster Y also points to M, In the absence of https://github.com/envoyproxy/envoy/issues/2943, you might be speaking to two different servers at the same time. You probably don't want update semantics in this case.

Reply:
Hmm.. How do we distinguish that case now? Do we have enough information or should we also include the cluster in the mix (that may be too much)?

Reply:
I think fundamentally you do need the cluster/listener there, or we need #2943 to land for EDS (as well as every other API that might have this property, e.g. RDS, SDS).

An alternative that might be simpler is making it a set (unordered list) of identifiers, but you need some way internally of knowing when to delete and modify the identifiers, so need some similar kind of association.